### DISCUSSION TITLE
Could anyone point me towards how the git extension works on a UI level?

### DISCUSSION THREAD

**Discussion by Savineerhs** [microsoft/vscode-discussions#297] [Extension Development QnA] [answered]
Hi!

Kinda new to the extension development scene, and I've recently implemented a lot of the core functionality of my extension. I'm looking into implementing a UI in the ViewContainer, and what I'm trying to achieve is a bit similar to the SQL Server/Git integration extension - show a few buttons/input fields for when the extension is "uninitialized" and then display a complex View. Problem is, I have no clue how to implement such a UI. I've spent a fair while staring at the git extension source code to not much avail either. 

For example, I noticed the git extension seems to use a TreeView in the "Changes" view but it also has an input field and a button above it? How does that work?

Thanks.
- Shree

**Comment by r3m0t** [comment, marked as answer]
The extension defines welcome content which can contain buttons, using the viewsWelcome contribution point. 

You basically have two options - use treeviews or a webview. A treeview can't contain text fields. A webview is like an iframe and you can style it to look "like" VS Code.

So how does the Git extension do it? There's two panels built into VS Code that get filled by extensions - Source Control and Testing. If you implement those APIs then your extension shows up in those panels. The UI elements are exactly what's designed in VS Code - Testing has a run and debug button and a filter text box, and Source Control has a commit message text box and a button. The APIs are exactly designed for tests or for version control systems, so if you use them for browsing SQL tables it'll confuse your users.

Here's the UI toolkit if you go for a webview: https://github.com/microsoft/vscode-webview-ui-toolkit

**Reply by gjsjohnmurray** [reply]
I'm not sure that the Testing view defaults to being in the SCM view container. More likely you dragged it there.

**Reply by Savineerhs** [reply]
> Source Control has a commit message text box and a button. 

So the Git extension is using a webview with a tree implemented in it as well?

**Reply by Savineerhs** [reply]
Oh wait, never mind. I think what you're saying is the Testing and Git extensions basically have UI code written specifically FOR them to behave the way they do, is that correct?

**Reply by r3m0t** [reply]
Yes SCM and Testing are VS Code features, Git and Mocha(eg) are extensions that use the features. The UI comes from VS Code.
